File: LeptonicaSharp/Extensions/BetterPictureBox.vb
Imports System.Windows.Forms
Imports System.Diagnostics
Imports System.Drawing
Imports System.IO


<DebuggerStepThrough()>
Public Class BetterPictureBox
    Inherits PictureBox

    Property SavedPic As Boolean = True
    Property LoadedPic As String = ""
    Property DrawSegments As Integer = 0
    Private _interpolationMode As Drawing2D.InterpolationMode = Drawing2D.InterpolationMode.Default

    Event ImageMouseClick(ByVal PT As MouseEventArgs)
    Event ImageMouseMove(ByVal PT As MouseEventArgs)
    Event ImageMouseUp(ByVal PT As MouseEventArgs)
    Event ImageMouseDown(ByVal PT As MouseEventArgs)

    Enum DrawMode
        Both
        Overlay
        Background
    End Enum
    Sub New()
        Me.BackColor = Color.LightGray
        Me.SizeMode = ImageLayout.Zoom
        Me.BorderStyle = Windows.Forms.BorderStyle.None
        Me.BackgroundImageLayout = ImageLayout.Zoom
    End Sub
    Public Property InterpolationMode As Drawing2D.InterpolationMode
        Get
            Return _interpolationMode
        End Get
        Set(ByVal value As Drawing2D.InterpolationMode)
            If value <> _interpolationMode Then
                _interpolationMode = value
                Me.Invalidate()
            End If
        End Set
    End Property

    <DebuggerStepThrough()>
    Protected Overrides Sub OnPaint(ByVal pe As System.Windows.Forms.PaintEventArgs)
        'pe.Graphics.InterpolationMode = _interpolationMode
        MyBase.OnPaint(pe)
    End Sub
    <DebuggerStepThrough()>
    Protected Overrides Sub OnPaintBackground(ByVal pevent As System.Windows.Forms.PaintEventArgs)
        pevent.Graphics.InterpolationMode = _interpolationMode
        MyBase.OnPaintBackground(pevent)
    End Sub

    <DebuggerStepThrough()>
    Function ZoomRand(ByVal PNT As Drawing.Point) As Drawing.Point
        If IsNothing(Me.BackgroundImage) Then Return PNT
        Dim ScaleW As Double = (Me.BackgroundImage.Width / Me.Width) '       Breitenskalierung
        Dim ScaleH As Double = (Me.BackgroundImage.Height / Me.Height) '     Hoehenskalierung
        Dim ScaleMin As Double = Math.Min(ScaleH, ScaleW) '                  Die Seite, die am kleinsten skaliert ist
        Dim ScaleMax As Double = Math.Max(ScaleH, ScaleW) '                  Die Seite, die am stärksten skaliert ist

        Dim DifPXW As Integer = Me.Width - (Me.BackgroundImage.Width / ScaleMax)    ' Anzahl der Randpixel Breite
        Dim DifPXH As Integer = Me.Height - (Me.BackgroundImage.Height / ScaleMax)  ' Anzahl der Randpixel Höhe

        Dim SolBreite As Integer = Me.Width
        Dim IstBreite As Integer = Me.BackgroundImage.Width / ScaleMax
        Dim DifBreite As Integer = (SolBreite - IstBreite) * ScaleMax

        Dim SolHoehe As Integer = Me.Height
        Dim IstHoehe As Integer = Me.BackgroundImage.Height / ScaleMax
        Dim DifHoehe As Integer = (SolHoehe - IstHoehe) * ScaleMax

        PNT.X = (PNT.X * ScaleMax) - DifBreite / 2
        PNT.Y = (PNT.Y * ScaleMax) - DifHoehe / 2
        Return PNT
    End Function

    <DebuggerStepThrough()>
    Function StretchRand(ByVal PNT As Drawing.Point) As Drawing.Point
        Dim ScaleW As Double = (Me.BackgroundImage.Width / Me.Width) '       Breitenskalierung
        Dim ScaleH As Double = (Me.BackgroundImage.Height / Me.Height) '     Hoehenskalierung
        Dim ScaleMin As Double = Math.Min(ScaleH, ScaleW)                       '    Die Seite, die am kleinsten skaliert ist
        Dim ScaleMax As Double = Math.Max(ScaleH, ScaleW)                        '   Die Seite, die am stärksten skaliert ist
        Dim ScaleDif As Double = ScaleMax - ScaleMin '                            '  Differenz der Faktoren
        PNT.X = (PNT.X * ScaleW)
        PNT.Y = (PNT.Y * ScaleH)
        Return PNT
    End Function

    Overloads Sub LoadImage(ByVal Picture As Image, Optional ByVal Overlay As Image = Nothing)
        If Not IsNothing(Me.Image) Then Me.Image.Dispose() : Me.Image = Nothing
        SavedPic = False

        If Me.Left < 0 Then Me.Left = 0
        If Me.Top < 0 Then Me.Top = 0
        If Me.Width < 100 Then Me.Width = 100
        If Me.Height < 100 Then Me.Height = 100

        If IsNothing(Picture) Then Exit Sub
        Me.BackgroundImage = Picture
        GenerateOverlay()
        RefreshPic()
    End Sub

    Delegate Sub PicRefresh()
    <DebuggerStepThrough()>
    Sub RefreshPic()
        Try
            If Me.InvokeRequired = True Then
                Me.Invoke(CType(AddressOf RefreshPic, PicRefresh))
                Exit Sub
            Else
                Me.Refresh()
            End If
        Catch : End Try
    End Sub

    Overloads Sub LoadImage(ByVal Filename As String)
        If Not IsNothing(Me.Image) Then Me.Image.Dispose() : Me.Image = Nothing
        If Not IsNothing(Me.BackgroundImage) Then Me.BackgroundImage.Dispose() : Me.BackgroundImage = Nothing
        LoadedPic = Filename
        SavedPic = True

        Dim Test As Boolean = False
        Dim TestI As Integer = 0
        Do Until Test = True Or TestI >= 20
            Try
                Dim FSTest As New System.IO.FileStream(LoadedPic, System.IO.FileMode.Open)
                FSTest.Dispose() : FSTest = Nothing : Test = True
                Using FS As New System.IO.FileStream(LoadedPic, System.IO.FileMode.Open)
                    Dim BMP As Bitmap = Bitmap.FromStream(FS) : Me.BackgroundImage = BMP
                    GenerateOverlay()
                End Using
            Catch
                TestI += 1
            End Try
        Loop

        RefreshPic()
    End Sub

    Function LoadUnblock(ByVal Filename As String) As Image
        Dim BMP As Bitmap
        If My.Computer.FileSystem.FileExists(Filename) = False Then Return Nothing
        Using FS As New System.IO.FileStream(Filename, System.IO.FileMode.Open)
            BMP = Bitmap.FromStream(FS)
        End Using
        Return BMP
    End Function

    <DebuggerStepThrough()>
    Sub GenerateOverlay()
        Try
            Dim OrgImage As Image = Me.BackgroundImage
            Dim Overlay As New Bitmap(OrgImage.Width, OrgImage.Height, Imaging.PixelFormat.Format32bppArgb)
            Overlay.SetResolution(OrgImage.HorizontalResolution, OrgImage.VerticalResolution)
        Catch : End Try
    End Sub

#Region "Picturebox Events + Zoom and Move"
    Dim MPosX As Integer
    Dim MPosY As Integer
    Enum ImageEvent
        MouseMove
        MouseClick
        MouseDown
        MouseUp
    End Enum

    <DebuggerStepThrough()>
    Sub SendEvent(ByVal e As MouseEventArgs, ByVal Evt As ImageEvent)
        Dim PNT As Drawing.Point = PointConverter(e)

        Select Case Evt
            Case ImageEvent.MouseClick : RaiseEvent ImageMouseClick(New MouseEventArgs(e.Button, e.Clicks, PNT.X, PNT.Y, e.Delta))
            Case ImageEvent.MouseDown : RaiseEvent ImageMouseDown(New MouseEventArgs(e.Button, e.Clicks, PNT.X, PNT.Y, e.Delta))
            Case ImageEvent.MouseMove : RaiseEvent ImageMouseMove(New MouseEventArgs(e.Button, e.Clicks, PNT.X, PNT.Y, e.Delta))
            Case ImageEvent.MouseUp : RaiseEvent ImageMouseUp(New MouseEventArgs(e.Button, e.Clicks, PNT.X, PNT.Y, e.Delta))
        End Select
    End Sub

    <DebuggerStepThrough()>
    Public Function PointConverter(ByVal Point As System.Windows.Forms.MouseEventArgs) As Drawing.Point
        Dim PNT As New Drawing.Point(Point.X, Point.Y)
        Select Case Me.SizeMode
            Case PictureBoxSizeMode.Zoom : PNT = ZoomRand(PNT)
            Case PictureBoxSizeMode.StretchImage : PNT = StretchRand(PNT)
        End Select
        Return PNT
    End Function

    <DebuggerStepThrough()>
    Private Sub PicBox_MouseWheel(ByVal sender As Object, ByVal e As System.Windows.Forms.MouseEventArgs) Handles Me.MouseWheel
        If e.Delta > 0 Then
            Me.Width = Me.Width + 200 : Me.Left = Me.Left - 100
            Me.Height = Me.Height + 200 : Me.Top = Me.Top - 100
        Else
            If Me.Height > 250 And Me.Width > 250 Then
                Me.Width = Me.Width - 200 : Me.Left = Me.Left + 100
                Me.Height = Me.Height - 200 : Me.Top = Me.Top + 100
            End If
        End If
        ' If Me.Left < Me.Width * -1 Then Me.Left = 0
    End Sub
    Private Sub PicBox_MouseDown(ByVal sender As Object, ByVal e As System.Windows.Forms.MouseEventArgs) Handles Me.MouseDown
        SendEvent(e, ImageEvent.MouseDown)
        If e.Button <> Windows.Forms.MouseButtons.Middle Then Exit Sub
        MPosX = e.X : MPosY = e.Y
    End Sub
    Private Sub PicBox_MouseUp(ByVal sender As Object, ByVal e As System.Windows.Forms.MouseEventArgs) Handles Me.MouseUp
        SendEvent(e, ImageEvent.MouseUp)
        If e.Button <> Windows.Forms.MouseButtons.Middle Then Exit Sub
        MPosX = 0 : MPosY = 0
    End Sub
    Private Sub PicBox_MouseClick(ByVal sender As Object, ByVal e As System.Windows.Forms.MouseEventArgs) Handles Me.MouseClick
        SendEvent(e, ImageEvent.MouseClick)
    End Sub
    <DebuggerStepThrough()>
    Private Sub PicBox_MouseMove(ByVal sender As Object, ByVal e As System.Windows.Forms.MouseEventArgs) Handles Me.MouseMove
        SendEvent(e, ImageEvent.MouseMove)
        Me.Select()
        If e.Button <> Windows.Forms.MouseButtons.Middle Then Exit Sub
        If MPosX = 0 Then Exit Sub
        If MPosY = 0 Then Exit Sub
        Me.Top = Me.Top - (MPosY - e.Y)
        Me.Left = Me.Left - (MPosX - e.X)
    End Sub
#End Region

End Class

File: CS_LIB/Extensions/BetterPictureBox.cs
using System.Collections.Generic;
using System.Collections;
using System.Windows.Forms;
using System.Diagnostics;
using System.Drawing;
using System.IO;
using System;

    [DebuggerStepThrough()]
    public class BetterPictureBox : PictureBox
    {
        private System.Drawing.Drawing2D.InterpolationMode _interpolationMode = System.Drawing.Drawing2D.InterpolationMode.NearestNeighbor;
        public event ImageMouseClickEventHandler ImageMouseClick;
        public delegate void ImageMouseClickEventHandler(MouseEventArgs PT);
        public event ImageMouseMoveEventHandler ImageMouseMove;
        public delegate void ImageMouseMoveEventHandler(MouseEventArgs PT);
        public event ImageMouseUpEventHandler ImageMouseUp;
        public delegate void ImageMouseUpEventHandler(MouseEventArgs PT);
        public event ImageMouseDownEventHandler ImageMouseDown;
        public delegate void ImageMouseDownEventHandler(MouseEventArgs PT);

        public enum DrawMode
        {
            Both,
            Overlay,
            Background
        }
        public BetterPictureBox()
        {
            MouseMove += PicBox_MouseMove;
            MouseClick += PicBox_MouseClick;
            MouseUp += PicBox_MouseUp;
            MouseDown += PicBox_MouseDown;
            MouseWheel += PicBox_MouseWheel;
            this.BackColor = Color.LightGray;
            this.SizeMode = PictureBoxSizeMode.Zoom;
            this.BorderStyle = System.Windows.Forms.BorderStyle.None;
            this.BackgroundImageLayout = ImageLayout.Zoom;
        }

        //public Drawing2D.InterpolationMode InterpolationMode
        //{
        //    get { return _interpolationMode; }
        //    set
        //    {
        //        if (value != _interpolationMode)
        //        {
        //            _interpolationMode = value;
        //            this.Invalidate();
        //        }
        //    }
        //}

        [DebuggerStepThrough()]
        protected override void OnPaint(System.Windows.Forms.PaintEventArgs pe)
        {
            //pe.Graphics.InterpolationMode = _interpolationMode
            base.OnPaint(pe);
        }
        [DebuggerStepThrough()]
        protected override void OnPaintBackground(System.Windows.Forms.PaintEventArgs pevent)
        {
            pevent.Graphics.InterpolationMode = _interpolationMode;
            base.OnPaintBackground(pevent);
        }

        //[DebuggerStepThrough()]
        //public System.Drawing.Point ZoomRand(System.Drawing.Point PNT)
        //{
        //    if ((this.BackgroundImage == null))
        //        return PNT;
        //    double ScaleW = (this.BackgroundImage.Width / this.Width);
        //    //       Breitenskalierung
        //    double ScaleH = (this.BackgroundImage.Height / this.Height);
        //    //     Hoehenskalierung
        //    double ScaleMin = Math.Min(ScaleH, ScaleW);
        //    //                  Die Seite, die am kleinsten skaliert ist
        //    double ScaleMax = Math.Max(ScaleH, ScaleW);
        //    //                  Die Seite, die am stärksten skaliert ist

        //    int DifPXW = this.Width - (this.BackgroundImage.Width / ScaleMax);
        //    // Anzahl der Randpixel Breite
        //    int DifPXH = this.Height - (this.BackgroundImage.Height / ScaleMax);
        //    // Anzahl der Randpixel Höhe

        //    int SolBreite = this.Width;
        //    int IstBreite = this.BackgroundImage.Width / ScaleMax;
        //    int DifBreite = (SolBreite - IstBreite) * ScaleMax;

        //    int SolHoehe = this.Height;
        //    int IstHoehe = this.BackgroundImage.Height / ScaleMax;
        //    int DifHoehe = (SolHoehe - IstHoehe) * ScaleMax;

        //    PNT.X = (PNT.X * ScaleMax) - DifBreite / 2;
        //    PNT.Y = (PNT.Y * ScaleMax) - DifHoehe / 2;
        //    return PNT;
        //}

        //[DebuggerStepThrough()]
        //public Drawing.Point StretchRand(System.Drawing.Point PNT)
        //{
        //    double ScaleW = (this.BackgroundImage.Width / this.Width);
        //    //       Breitenskalierung
        //    double ScaleH = (this.BackgroundImage.Height / this.Height);
        //    //     Hoehenskalierung
        //    double ScaleMin = Math.Min(ScaleH, ScaleW);
        //    //    Die Seite, die am kleinsten skaliert ist
        //    double ScaleMax = Math.Max(ScaleH, ScaleW);
        //    //   Die Seite, die am stärksten skaliert ist
        //    double ScaleDif = ScaleMax - ScaleMin;
        //    //                            '  Differenz der Faktoren
        //    PNT.X = ( PNT.X * (int)ScaleW);
        //    PNT.Y = (PNT.Y * (int)ScaleH);
        //    return PNT;
        //}

        public void LoadImage(Image Picture, Image Overlay = null)
        {
            if ((this.Image != null)) { this.Image.Dispose(); this.Image = null; }

            if (this.Left < 0)
                this.Left = 0;
            if (this.Top < 0)
                this.Top = 0;
            if (this.Width < 100)
                this.Width = 100;
            if (this.Height < 100)
                this.Height = 100;

            if ((Picture == null))
                return;
            this.BackgroundImage = Picture;
            RefreshPic();
        }

        public delegate void PicRefresh();
        [DebuggerStepThrough()]
        public void RefreshPic()
        {
            try
            {
                if (this.InvokeRequired == true)
                {
                    this.Invoke((PicRefresh)RefreshPic);
                    return;
                }
                else
                {
                    this.Refresh();
                }
            }
            catch
            {
            }
        }

        public void LoadImage(string Filename)
        {
            if ((this.Image != null)) { this.Image.Dispose(); this.Image = null; }
            if ((this.BackgroundImage != null)) { this.BackgroundImage.Dispose(); this.BackgroundImage = null; }

            bool Test = false;
            int TestI = 0;
            while (!(Test == true | TestI >= 20))
            {
                try
                {
                    System.IO.FileStream FSTest = new System.IO.FileStream(Filename, System.IO.FileMode.Open);
                    FSTest.Dispose();
                    FSTest = null;
                    Test = true;
                    using (System.IO.FileStream FS = new System.IO.FileStream(Filename, System.IO.FileMode.Open))
                    {
                        //Bitmap BMP = Bitmap.FromStream(FS);
                        //this.BackgroundImage = BMP;
                    }
                }
                catch
                {
                    TestI += 1;
                }
            }

            RefreshPic();
        }

        public Image LoadUnblock(string Filename)
        {
            Bitmap BMP = null;
            if (File.Exists(Filename) == false)
                return null;
            using (System.IO.FileStream FS = new System.IO.FileStream(Filename, System.IO.FileMode.Open))
            {
                // BMP = Bitmap.FromStream(FS);
            }
            return BMP;
        }

        #region "Picturebox Events + Zoom and Move"
        int MPosX;
        int MPosY;
        public enum ImageEvent
        {
            MouseMove,
            MouseClick,
            MouseDown,
            MouseUp
        }

        [DebuggerStepThrough()]
        public void SendEvent(MouseEventArgs e, ImageEvent Evt)
        {
            System.Drawing.Point PNT= PointConverter(e);

            switch (Evt)
            {
                case ImageEvent.MouseClick:
                    if (ImageMouseClick != null)
                    {
                        ImageMouseClick(new MouseEventArgs(e.Button, e.Clicks, PNT.X, PNT.Y, e.Delta));
                    }

                    break;
                case ImageEvent.MouseDown:
                    if (ImageMouseDown != null)
                    {
                        ImageMouseDown(new MouseEventArgs(e.Button, e.Clicks, PNT.X, PNT.Y, e.Delta));
                    }

                    break;
                case ImageEvent.MouseMove:
                    if (ImageMouseMove != null)
                    {
                        ImageMouseMove(new MouseEventArgs(e.Button, e.Clicks, PNT.X, PNT.Y, e.Delta));
                    }

                    break;
                case ImageEvent.MouseUp:
                    if (ImageMouseUp != null)
                    {
                        ImageMouseUp(new MouseEventArgs(e.Button, e.Clicks, PNT.X, PNT.Y, e.Delta));
                    }

                    break;
            }
        }

        [DebuggerStepThrough()]
        public System.Drawing.Point PointConverter(System.Windows.Forms.MouseEventArgs Point)
        {
            //System.Drawing.Point PNT = new System.Drawing.Point(Point.X, Point.Y);
            //switch (this.SizeMode)
            //{
            //    case PictureBoxSizeMode.Zoom:
            //        PNT = ZoomRand(PNT);
            //        break;
            //    case PictureBoxSizeMode.StretchImage:
            //        PNT = StretchRand(PNT);
            //        break;
            //}
            //return PNT;
            return new System.Drawing.Point (0,0);
        }

        private void PicBox_MouseDown(object sender, System.Windows.Forms.MouseEventArgs e)
        {
            SendEvent(e, ImageEvent.MouseDown);
            if (e.Button != System.Windows.Forms.MouseButtons.Middle)
                return;
            MPosX = e.X;
            MPosY = e.Y;
        }
        private void PicBox_MouseUp(object sender, System.Windows.Forms.MouseEventArgs e)
        {
            SendEvent(e, ImageEvent.MouseUp);
            if (e.Button != System.Windows.Forms.MouseButtons.Middle)
                return;
            MPosX = 0;
            MPosY = 0;
        }
        private void PicBox_MouseClick(object sender, System.Windows.Forms.MouseEventArgs e)
        {
            SendEvent(e, ImageEvent.MouseClick);
        }

        [DebuggerStepThrough()]
        private void PicBox_MouseWheel(object sender, System.Windows.Forms.MouseEventArgs e)
        {
            if (e.Delta > 0)
            {
                this.Width = this.Width + 200;
                this.Left = this.Left - 100;
                this.Height = this.Height + 200;
                this.Top = this.Top - 100;
            }
            else
            {
                if (this.Height > 250 & this.Width > 250)
                {
                    this.Width = this.Width - 200;
                    this.Left = this.Left + 100;
                    this.Height = this.Height - 200;
                    this.Top = this.Top + 100;
                }
            }
            // If Me.Left < Me.Width * -1 Then Me.Left = 0
        }
        [DebuggerStepThrough()]
        private void PicBox_MouseMove(object sender, System.Windows.Forms.MouseEventArgs e)
        {
            SendEvent(e, ImageEvent.MouseMove);
            this.Select();
            if (e.Button != System.Windows.Forms.MouseButtons.Middle)
                return;
            if (MPosX == 0)
                return;
            if (MPosY == 0)
                return;
            this.Top = this.Top - (MPosY - e.Y);
            this.Left = this.Left - (MPosX - e.X);
        }
        #endregion

    }


